### DISCUSSION TITLE
indexOf

### DISCUSSION THREAD

**Discussion by prdoyle** [hrldcpr/pcollections#110] [Ideas]
In my application, it would be nice to look up the position of a given key in an `OrderedPMap`; that is, how many elements are to the left of a given element. This should be straightforward to implement in log(n) time in `KVTree`, and exposed in `OrderedPMap`, `OrderedPSet`, `PSortedMap`, and `PSortedSet`.

If you're open to this idea, I could start working on an implementation.

A natural follow-up question would be: should we be able to fetch an elemenet given its index? I don't need that myself, but I could add it if we want it and it's straightforward enough. (Though I might need some help picking a method name; maybe `getAt`?)

**Comment by prdoyle** [comment]
I suppose `KVTree.rangeToLeft(key, false, comparator).size` already has the right semantics, though it creates a number of objects that could be avoided with a purpose-built `KVTree.indexOf` implementation.

**Comment by prdoyle** [comment]
Branch under development [here](https://github.com/prdoyle/pcollections/tree/indexof). I picked semantics consistent with `Arrays.binarySearch` for the case where the element is not present.

**Comment by hrldcpr** [comment]
sorry for the slow response, sounds cool to me!
possibly @ran-arigur might have some insights/input since he wrote that code

**Comment by prdoyle** [comment]
No sweat, I got distracted by something shiny and haven't worked on this for a while.

**Comment by ran-arigur** [comment]
Makes sense to me! :slightly_smiling_face: 

For the map implementations, though, I think the method should be called `indexOfKey(...)` rather than `indexOf(...)`. (Cf. `containsKey(...)`, `firstKey()`, etc.) And their Javadoc should refer to "keys" or "mappings" rather than "elements".

Aside from that, a few small thoughts for your consideration:

- Since OrderedPSet and OrderedPMap are unsorted, I think it would make sense to use the simpler semantics of List.indexOf(...) for those two, rather than the full semantics of Arrays.binarySearch(...).
    - Relatedly, the `@see` links between the 'ordered' and 'sorted' implementations of `indexOf(...)` are more confusing than helpful IMHO.
- Adding an abstract method to an interface is a backward-incompatible change if there are any outside implementations of that interface &mdash; which seems unlikely here, but who knows! If we're concerned about that, then maybe PSortedSet and PSortedMap can provide default implementations that use headSet(...)/headMap(...), last()/lastKey(), and size()?
- It would be better not to introduce the word "left" into the Javadoc, if we can help it, since the premise that "to the left of" = "before" may depend on use-case, cultural context, etc.
    - For OrderedPSet, maybe something like "the number of elements of this set that precede `element` in iteration order, if `element` is in this set"? Or maybe something like "the index of `element`, if it is in this set: `0` if it is the first element, `size-1` if it is the last"? (And likewise for OrderedPMap.)
    - For PSortedSet, maybe something like "the number of elements of this set that precede (are less than) `element`, if `element` is in this set"? Or maybe something like "the index of `element`, if it is in this set: `0` if it is the first (lowest) element, `size-1` if it is the last (highest)"? (And likewise for PSortedMap.)
- I think it would be helpful to copy this observation from the Javadoc for Arrays.binarySearch: "Note that this guarantees that the return value will be >= 0 if and only if the key is found." (Just because the whole `(-n - 1)` thing is so tricky IMHO that it's worth it to avoid people having to reason through it in cases where the details don't matter to them.) Alternatively, instead of "otherwise, `(-n - 1)` where n is [&hellip;]", maybe something like "otherwise, a negative number *n* such that `-n-1` is the number of elements less than `element`"?

But those are all minor things. `indexOfKey(...)` is the only thing I feel pretty strongly about. :slightly_smiling_face:

> A natural follow-up question would be: should we be able to fetch an elemenet given its index? I don't need that myself, but I could add it if we want it and it's straightforward enough. (Though I might need some help picking a method name; maybe `getAt`?)

That makes sense! Maybe `getAtIndex` for the sets, and `getKeyAtIndex` and `getEntryAtIndex` for the maps?

**Comment by ran-arigur** [comment]
Actually, wait, sorry, there's a bug: `TreePSet.indexOf(...)` and `TreePMap.indexOf(...)` don't correctly handle the "descending" case, where `this.isLeftToRight` is `false` and `this.comparator()` is the reverse of `this.ltrComparator`. We can't pass `this.comparator()` into `this.tree.indexOf(...)`, because the latter assumes that the comparator that's passed in matches the order of mappings in the tree; for example, `TreePSet.from(Arrays.asList(0, 1, 2, 3, 4, 5, 6, 7)).descendingSet().indexOf(2)` will return a negative value, because KVTree.indexOf(...) will look for `2` on the wrong side of the tree.

To fix this, we need to explicitly consider `this.isLeftToRight`; for example, something like this:

```
@Override
public int indexOf(final E e) {
  final int leftToRightIndex = this.tree.indexOf(e, this.ltrComparator);
  if (this.isLeftToRight) {
    return leftToRightIndex;
  } else if (leftToRightIndex >= 0) {
    return this.size() - leftToRightIndex - 1;
  } else {
    return -this.size() - leftToRightIndex - 2;
  }
} 
```

(Disclaimer: not tested.)

Instead of having tests for the internal `KVTree.indexOf(...)` method, I think it would be better to have tests for the client-facing API methods (`TreePSet.indexOf(...)`, etc.); that's a better practice in general, and in this particular case it lets us cover both the "ascending" and "descending" cases. Otherwise it's just too easy otherwise to mishandle the "descending" case without realizing!

**Reply by prdoyle** [reply]
Ah! I'm still getting used to pcollections and missed that `KVTree` isn't public.

Thanks again for the tips!

**Comment by prdoyle** [comment]
Amazing, thanks so much for these thoughtful reviews! I haven't had time to look at this for a few weeks but hopefully soon.

**Comment by prdoyle** [comment]
Hey I just realized PR #98 removed `indexOf` with the following comment:

> This change means get() and indexOf() can't be efficient anymore, but they should probably have never been there in the first place, since the analogous class LinkedHashSet doesn't have them, so this removes them.

Am I naively putting back functionality that is already known to be a bad idea?

**Reply by hrldcpr** [reply]
I think the main reason I removed them is because the naive implementation would've been linear time, but now that they can be faster I'm down to have them again!

**Comment by prdoyle** [comment]
Still hoping to get some time to look at this soon...

**Comment by ran-arigur** [comment]
@prdoyle If you don't have time for this soon, would you be offended and/or disappointed if I were to take it over and update the PR per my comments above? (Note: I myself won't be able to get to this for at least the next few weeks, so "soon" is relative. No rush!)

**Reply by prdoyle** [reply]
I would not. Go for it!

**Comment by ran-arigur** [comment]
Sorry for the delay here. I tried to work on this about a month and a half ago, but I found that, between Gradle settings and Java 9 modules and ParameterizedTests, it's become quite difficult to develop PCollections in Eclipse; I had to do a bunch of weird Eclipse setup, and even after that, there were still major problems (it couldn't resolve the imports in Benchmarks.java, and it couldn't handle the unit-tests that use ParameterizedTests).

So, I assume that y'all are using a different IDE? It would be helpful if there were a bit of documentation somewhere about how to get set up &mdash; even if it's just a line saying "We use [IDE name], see [link to its documentation]" &mdash; unless there already is and I just haven't found it. (I notice that \*.iml is mentioned in .gitignore, which suggests that at least one contributor used IntelliJ IDEA at some point; is that what you use today?)

(Also, by the way, on the subject of new-user setup: This is on me, but I didn't realize until just now that I could address Gradle's complaint about `Could not get unknown property 'ossrhUsername' for Credentials [username: null] of type org.gradle.internal.credentials.DefaultPasswordCredentials_Decorated` by adding a gradle.properties file with fake values of ossrhUsername and ossrhPassword. I *did* see the documentation at https://github.com/hrldcpr/pcollections/blob/master/CHANGING.md, but since most of that doc is directed at someone who's actually releasing changes, it wasn't obvious to me that I could fake-do the first part without having real credentials. Would you accept a pull request that just added a gradle.properties with fake credentials? I believe it's possible to do that while keeping it .gitignore-d so that people who *do* have real credentials can put them in their gradle.properties without Git wanting to track their changes to it.)

**Reply by hrldcpr** [reply]
Would love a pull request with fake credentials! In fact, #112 happened to be opened in the past week with the same exact issue..

**Reply by hrldcpr** [reply]
As for IDE, I actually just use a relatively vanilla emacs for everything (which works okay since I rarely use Java), so just running `gradlew` seems to work for me, but would love any pull requests that make a more legitimate workflow doable.

**Comment by prdoyle** [comment]
I use IntelliJ FWIW.
